File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void findcombination(int ind, int target, vector<int>&nums,
                         vector<vector<int>>&ans, vector<int>&ds) {
        if (ind == nums.size()) {
            if (target == 0) {
                ans.push_back(ds);
            }
            return;
        }

        if (nums[ind] <= target) {
            ds.push_back(nums[ind]);
            findcombination(ind, target - nums[ind], nums, ans, ds);
            ds.pop_back();
        }
        findcombination(ind + 1, target, nums, ans, ds);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>ans;
        vector<int>ds;
        findcombination(0,target,candidates,ans,ds);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    void findcombination(int ind,int target,int[] candidates,List<List<Integer>>ans,List<Integer> ds){
        if(ind == candidates.length){
            if(target == 0){
                ans.add(new ArrayList<>(ds));
            }
            return;
        }
        if(candidates[ind] <= target){
            ds.add(candidates[ind]);
            findcombination(ind ,target-candidates[ind],candidates,ans,ds);
            ds.remove(ds.size()-1);
        }
        findcombination(ind+1,target,candidates,ans,ds);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>ans = new ArrayList<>();
        List<Integer>ds = new ArrayList<>();

        findcombination(0,target,candidates,ans,ds);
        return ans;
    }
}